core/src/error.rs

```diff
@@ -0,0 +1,212 @@
+// documentation files (the "Software"), to deal in the
+// Software without restriction, including without
+// limitation the rights to use, copy, modify, merge,
+// publish, distribute, sublicense, and/or sell copies of
+// the Software, and to permit persons to whom the Software
+// is furnished to do so, subject to the following
+// conditions:
+//
+// The above copyright notice and this permission notice
+// shall be included in all copies or substantial portions
+// of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF
+// ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED
+// TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
+// PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT
+// SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
+// CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+// OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR
+// IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+// DEALINGS IN THE SOFTWARE.
+
+use std::fmt;
+
+use jsonrpsee_types::error::CallError;
+use serde::{Deserialize, Serialize};
+
+/// Convenience type for displaying errors.
+#[derive(Clone, Debug, PartialEq)]
+pub struct Mismatch<T> {
+	/// Expected value.
+	pub expected: T,
+	/// Actual value.
+	pub got: T,
+}
+
+impl<T: fmt::Display> fmt::Display for Mismatch<T> {
+	fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+		f.write_fmt(format_args!("Expected: {}, Got: {}", self.expected, self.got))
+	}
+}
+
+// NOTE(niklasad1): this `From` impl is a bit opinionated to regard all generic errors as `CallError`.
+// In practice this should be the most common use case for users of this library.
+impl From<anyhow::Error> for Error {
+	fn from(err: anyhow::Error) -> Self {
+		Error::Call(CallError::Failed(err))
+	}
+}
+
+/// Error type.
+#[derive(Debug, thiserror::Error)]
+pub enum Error {
+	/// Error that occurs when a call failed.
+	#[error("Server call failed: {0}")]
+	Call(#[from] CallError),
+	/// Networking error or error on the low-level protocol layer.
+	#[error("Networking or low-level protocol error: {0}")]
+	Transport(#[source] anyhow::Error),
+	/// JSON-RPC request error.
+	#[error("JSON-RPC request error: {0:?}")]
+	Request(String),
+	/// Frontend/backend channel error.
+	#[error("Frontend/backend channel error: {0}")]
+	Internal(#[from] futures_channel::mpsc::SendError),
+	/// Invalid response,
+	#[error("Invalid response: {0}")]
+	InvalidResponse(Mismatch<String>),
+	/// The background task has been terminated.
+	#[error("The background task been terminated because: {0}; restart required")]
+	RestartNeeded(String),
+	/// Failed to parse the data.
+	#[error("Parse error: {0}")]
+	ParseError(#[from] serde_json::Error),
+	/// Invalid subscription ID.
+	#[error("Invalid subscription ID")]
+	InvalidSubscriptionId,
+	/// Invalid request ID.
+	#[error("Invalid request ID")]
+	InvalidRequestId,
+	/// Client received a notification with an unregistered method
```

Comment:
```suggestion
	/// Client received a notification with an unregistered method.
```

Reply:
I don't want to fill the PR with unnecessary noise, so I just made one suggestion, but not sure if you want the docs to all have periods after each description. Just an observation

Reply:
Yea makes sense :). Will close this one.